x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -0,0 +1,386 @@
+  template: {
+    index_patterns: SESSION_INDEX_NAME,
+    order: 1000,
+    settings: {
+      index: {
+        number_of_shards: 1,
+        number_of_replicas: 0,
+        auto_expand_replicas: '0-1',
+        priority: 1000,
+        refresh_interval: '1s',
+        hidden: true,
+      },
+    },
+    mappings: {
+      dynamic: 'strict',
+      properties: {
+        username_hash: { type: 'keyword' },
+        provider: { properties: { name: { type: 'keyword' }, type: { type: 'keyword' } } },
+        path: { type: 'keyword' },
+        idleTimeoutExpiration: { type: 'date' },
+        lifespanExpiration: { type: 'date' },
+        accessAgreementAcknowledged: { type: 'boolean' },
+        content: { type: 'binary' },
+      },
+    },
+    aliases: { [SESSION_INDEX_ALIAS]: {} },
+  },
+};
+
+/**
+ * Represents shape of the session value stored in the index.
+ */
+export interface SessionIndexValue {
+  /**
+   * Unique session ID.
+   */
+  sid: string;
+
+  /**
+   * Hash of the username. It's defined only if session is authenticated, otherwise session
+   * is considered unauthenticated (e.g. intermediate session used during SSO handshake).
+   */
+  username_hash?: string;
+
+  /**
+   * Name and type of the provider this session belongs to.
+   */
+  provider: AuthenticationProvider;
+
+  /**
+   * The Unix time in ms when the session should be considered expired. If `null`, session will stay
+   * active until the browser is closed.
+   */
+  idleTimeoutExpiration: number | null;
+
+  /**
+   * The Unix time in ms which is the max total lifespan of the session. If `null`, session expire
+   * time can be extended indefinitely.
+   */
+  lifespanExpiration: number | null;
+
+  /**
+   * Kibana server base path the session was created for.
+   */
+  path: string;
+
+  /**
+   * Indicates whether user acknowledged access agreement or not.
+   */
+  accessAgreementAcknowledged?: boolean;
+
+  /**
+   * Content of the session value represented as an encrypted JSON string.
+   */
+  content: string;
+
+  /**
+   * Additional index specific information about the session value.
+   */
+  metadata: SessionIndexValueMetadata;
+}
+
+/**
+ * Additional index specific information about the session value.
+ */
+interface SessionIndexValueMetadata {
+  /**
+   * Primary term of the last modification of the document.
+   */
+  primaryTerm: number;
+
+  /**
+   * Sequence number of the last modification of the document.
+   */
+  sequenceNumber: number;
+}
+
+export class SessionIndex {
+  /**
+   * Timeout after which session with the expired idle timeout _may_ be removed from the index
+   * during regular cleanup routine. It's intentionally larger than `idleIndexUpdateTimeout`
+   * configured in `Session` to be sure that the session value may be safely cleaned up.
+   */
+  readonly #idleIndexCleanupTimeout: number | null;
+
+  /**
+   * Options used to create Session index.
+   */
+  readonly #options: Readonly<SessionIndexOptions>;
+
+  /**
```

Comment:
__note:__ I don't like this and to be honest our test runner was that last straw that forced me to do that - we don't have any way yet to say that plugin is ready and can accept requests. Our FTR doesn't wait until we initialize index and tries to log in and fails from time to time. So added this just for the initial bootstrap, in real scenarios it sounds OK if login attempt fails when Kibana just starting up, but for tests it's critical. And it's not possible to account for that in just one test suite - almost every functional test relies on login. I plan to get rid of this as soon as we have a proper Core status service (https://github.com/elastic/kibana/issues/41983) that will allow us to indicate our status. And once we have it FTR will wait for overall green status before it can consider Kibana as ready.

Reply:
Yep, that's a good idea, will do!